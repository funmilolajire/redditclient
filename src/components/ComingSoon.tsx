import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';

export const ComingSoon = () => {
    const boxRef = useRef<HTMLDivElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({ repeatDelay: 2 })
        tl.from([boxRef.current], {
            y: '-55vh',
            duration: 3,
            delay: 2,
            opacity: 0,
            ease: 'elastic'
        })
    })

    return (
        <div ref={boxRef} className="container flex items-center justify-center h-screen font-extrabold ComingSoon">
            <div className="bg-yellow-500 border-2 border-gray-800 border-solid skew-y-2 transform-gpu text-gray-50 text-9xl">
                <p>Coming Soon</p>
            </div>
        </div>
    )
}