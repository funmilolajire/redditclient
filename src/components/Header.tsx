import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';
import logo from '../images/logo192.png';

export const Header = () => {
    const boxRef = useRef<HTMLDivElement>(null);
    const logoRef = useRef<HTMLImageElement>(null);
    const titleRef = useRef<HTMLHeadingElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({});
        tl.from([boxRef.current], {
            duration: 1
        })
        tl.from([logoRef.current], {
            x: '-3000px',
            opacity: 0,
            ease: 'circ'
        })
        tl.from([titleRef.current], {
            x: '3000px',
            opacity: 0,
            ease: 'slow'
        })
    })

    return (
        <header>
            <div className="header-inner">
                <section ref={boxRef}>
                    <figure ref={logoRef}>
                        <img src={logo} alt="site logo" />
                    </figure>
                    <h1 ref={titleRef}>
                        <span>Reddit</span>Client
                    </h1>
                </section>
            </div>
        </header>
    )
}