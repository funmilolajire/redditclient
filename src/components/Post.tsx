import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';

interface PostType {
    post: {
        likes: number;
        comments: number;
        postTitle: string;
        user: string;
        subreddit: string;
        time: string;
        postImage: string;
    }
}

export const Post = (props: PostType) => {
    const { likes, comments, postTitle, user, subreddit, time, postImage } = props.post;

    const boxRef = useRef<HTMLDivElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([boxRef.current], {
            y: '150%',
            duration: 2,
            delay: 3,
            ease: 'back'
        })
    }, [])

    return (
        <div className="Post">
            <div ref={boxRef} className="Post--Container ">
                <div className=" Post-Likes">
                    <p><span className="likes">{likes}</span><span className="label">likes</span></p>
                </div>
                <div className="Post--Content">
                    <figure>
                        <img src={postImage} alt="post" />
                    </figure>
                    <section>
                        <h3 className="postTitle">{postTitle}</h3>
                        <p>submitted <span className="time">{time}</span> by <span className="user">{user}</span> to <span className="subreddit">{subreddit}</span></p>
                        <div className="Post--Comments"><span className="comments">{comments}</span> comments</div>
                    </section>
                </div>
            </div>
        </div>
    )
}
Post.defaultProps = {
    post: {
        likes: 0,
        postTitle: '',
        postImage: '',
        time: '',
        user: '',
        subreddit: '',
        comments: 0
    }
}