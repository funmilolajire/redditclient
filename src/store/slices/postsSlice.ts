import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { searchResults, subredditPosts } from '../api';

//set initial state
const initialState: PostsState = {
    posts: [],
    loading: true,
    listing: 'hot',
    term: '',
    subreddit: ''
}

//get subreddit,listing posts and search results
export const getSubredditPosts = createAsyncThunk('subredditPosts', subredditPosts)
export const getSearchResults = createAsyncThunk('searchResults', searchResults)

//create posts slice
const postsSlice = createSlice({
    name: "subreddits",
    initialState,
    reducers: {
        handleListingChange: (state: PostsState, { payload }: { payload: string }) => {
            state.loading = true;
            state.listing = payload
        },
        handleSubredditChange: (state: PostsState, { payload }: { payload: string }) => {
            state.loading = true;
            state.subreddit = payload
        },
        handleTermChange: (state: PostsState, { payload }: { payload: string }) => {
            state.term = payload;
        },
        //for testing
        handleLoadingChange: (state: PostsState, { payload }: { payload: boolean }) => {
            state.loading = payload;
        }
    },
    extraReducers: builder => {
        builder.addCase(getSubredditPosts.fulfilled, (state: PostsState, { payload, meta }) => {
            if (payload)
                state.posts = payload;
            state.loading = false;
            if (meta.arg.listing) {
                state.listing = meta.arg.listing.substring(1);
            }
            if (meta.arg.subreddit) {
                state.subreddit = meta.arg.subreddit.substring(1);
            }
        });
        builder.addCase(getSearchResults.fulfilled, (state: PostsState, { payload }) => {
            if (payload)
                state.posts = payload;
            state.loading = false;
        })
    }
})

//export posts slice reducer and actions
export default postsSlice.reducer;
export const { handleListingChange, handleSubredditChange, handleTermChange, handleLoadingChange } = postsSlice.actions;