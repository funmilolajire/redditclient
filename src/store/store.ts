import { configureStore } from '@reduxjs/toolkit';
import termReducer from './slices/termSlice';

export default configureStore({
  reducer: {
    term: termReducer,
  },
});
