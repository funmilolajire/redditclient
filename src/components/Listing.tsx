import { useRef, useLayoutEffect, useCallback } from 'react';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { handleListingChange, handleTermChange } from '../store/slices/postsSlice';

export const Listing = ({ listItem }: { listItem: string }) => {
    const listingRef = useRef<HTMLLIElement>(null);

    //handle listing change and active class
    const dispatch = useAppDispatch();
    const listing = useAppSelector(state => state.posts.listing);
    const handleListing = () => {
        if (null !== listingRef.current && listing !== listingRef.current.id) {
            let newListing = listingRef.current.id;
            dispatch(handleListingChange(newListing))
            dispatch(handleTermChange(''))
        }
    }
    const handleActiveClass = useCallback(() => {
        if (null !== listingRef.current) {
            const classList = listing === listingRef.current.id ? "Listing active" : "Listing";
            listingRef.current.className = classList
        }
    }, [listing])
    useLayoutEffect(() => {
        handleActiveClass();
    }, [handleActiveClass])


    //JSX
    return (
        <li ref={listingRef} data-testid={listItem} onClick={handleListing} id={listItem}>
            {listItem}
        </li>
    )
}