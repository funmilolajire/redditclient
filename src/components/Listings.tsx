import { useRef, useLayoutEffect } from 'react';
import { Listing } from './Listing';
import { gsap } from 'gsap';

export const Listings = () => {
    const listings: string[] = ['hot', 'new', 'top', 'best', 'rising', 'controversial'];

    //animations
    const listingBoxRef = useRef<HTMLDivElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([listingBoxRef.current], {
            x: '1200%',
            delay: 2,
            duration: 2,
            ease: 'power3.inOut'
        })
    }, [])

    return (
        <div ref={listingBoxRef} className="Listings">
            <ul className="Listings--List">
                {
                    listings && listings.map(listing => <Listing key={listing} listItem={listing} />)
                }
            </ul>
        </div>
    )
}