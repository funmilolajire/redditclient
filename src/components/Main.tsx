import { Listings } from './Listings';
import { Posts } from './Posts';
export const Main = () => (
    <div className="Main">
        <Listings />
        <Posts />
    </div>
)