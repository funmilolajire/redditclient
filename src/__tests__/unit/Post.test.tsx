import { mount } from 'enzyme';
import { Post } from '../../components/Post';
import { Provider } from 'react-redux';
import { store } from '../../store/store';
import { handleLoadingChange } from '../../store/slices/postsSlice';


describe('renders Post component', () => {
    store.dispatch(handleLoadingChange(false))
    const props = {
        post: {
            "comments": 10,
            "downvotes": 0,
            "postAltIframe": "",
            "postAltVideo": "",
            "postIframe": "",
            "postImage": "https://b.thumbs.redditmedia.com/5ECi_D2EZxcveWeDzFK8bdMu2pXXeu_oO-cxk0AEtYs.jpg",
            "postThumbnail": "",
            "postVideo": "",
            "permalink": "/r/wallstreetbets/comments/mi3t1v/weekend_discussion_thread_for_the_weekend_of/",
            "postBody": "&lt;!-- SC_OFF --&gt;&lt;div class=\"md\"&gt;&lt;p&gt;Your weekend discussion thread. Please keep the shitposting to a maximum!&lt;/p&gt;\n\n&lt;p&gt;Follow &lt;a href=\"https://twitter.com/Official_WSB\"&gt;@Official_WSB&lt;/a&gt; on Twitter, all other accounts are impersonators.&lt;/p&gt;\n&lt;/div&gt;&lt;!-- SC_ON --&gt;",
            "postTitle": "Weekend Discussion Thread for the Weekend of April 01, 2021",
            "subreddit": "r/wallstreetbets",
            "time": 1617307633.0,
            "upvotes": 10,
            "user": "OPINION_IS_UNPOPULAR",
        }
    }
    const PostComponent = mount(
        <Provider store={store}>
            <Post {...props} />
        </Provider>
    )

    test('make snapshot of Post component', () => {
        expect(PostComponent).toMatchSnapshot()
    })

    // test('modal opens on comment click', () => {
    //     const comment = PostComponent.find('.Post .Post--Container .Post--Content--Wrapper .Post--Content .Post--Comments').first();
    //     comment.simulate('click');
    //     const showModal = store.getState().moreInformation.showModal;
    //     expect(showModal).toBeTrue();
    // })

    test('changes current subreddit to clicked subreddit', () => {
        const subredditChange = PostComponent.find('.Post--Container .subreddit');
        subredditChange.simulate('click')
        const subreddit = store.getState().posts.subreddit
        expect(subreddit).toBe(subredditChange.text());
    })
})