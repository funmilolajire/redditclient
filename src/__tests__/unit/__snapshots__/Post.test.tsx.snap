// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Post component make snapshot of Post component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Post
    post={
      Object {
        "comments": 10,
        "downvotes": 0,
        "permalink": "/r/wallstreetbets/comments/mi3t1v/weekend_discussion_thread_for_the_weekend_of/",
        "postAltIframe": "",
        "postAltVideo": "",
        "postBody": "&lt;!-- SC_OFF --&gt;&lt;div class=\\"md\\"&gt;&lt;p&gt;Your weekend discussion thread. Please keep the shitposting to a maximum!&lt;/p&gt;

&lt;p&gt;Follow &lt;a href=\\"https://twitter.com/Official_WSB\\"&gt;@Official_WSB&lt;/a&gt; on Twitter, all other accounts are impersonators.&lt;/p&gt;
&lt;/div&gt;&lt;!-- SC_ON --&gt;",
        "postIframe": "",
        "postImage": "https://b.thumbs.redditmedia.com/5ECi_D2EZxcveWeDzFK8bdMu2pXXeu_oO-cxk0AEtYs.jpg",
        "postThumbnail": "",
        "postTitle": "Weekend Discussion Thread for the Weekend of April 01, 2021",
        "postVideo": "",
        "subreddit": "r/wallstreetbets",
        "time": 1617307633,
        "upvotes": 10,
        "user": "OPINION_IS_UNPOPULAR",
      }
    }
  >
    <div
      className="Post"
      data-permalink="/r/wallstreetbets/comments/mi3t1v/weekend_discussion_thread_for_the_weekend_of/"
    >
      <div
        className="Post--Container"
      >
        <p
          className="Post--Info"
        >
          posted 
          <span
            className="time"
          >
            1617307633
          </span>
           by 
          <span
            className="user"
          >
            OPINION_IS_UNPOPULAR
          </span>
           to 
          <span
            className="subreddit"
            data-subreddit="r/wallstreetbets"
            onClick={[Function]}
          >
            r/wallstreetbets
          </span>
        </p>
        <div
          className="Post--Content--Wrapper"
        >
          <PostContent
            post={
              Object {
                "comments": 10,
                "downvotes": 0,
                "permalink": "/r/wallstreetbets/comments/mi3t1v/weekend_discussion_thread_for_the_weekend_of/",
                "postAltIframe": "",
                "postAltVideo": "",
                "postBody": "&lt;!-- SC_OFF --&gt;&lt;div class=\\"md\\"&gt;&lt;p&gt;Your weekend discussion thread. Please keep the shitposting to a maximum!&lt;/p&gt;

&lt;p&gt;Follow &lt;a href=\\"https://twitter.com/Official_WSB\\"&gt;@Official_WSB&lt;/a&gt; on Twitter, all other accounts are impersonators.&lt;/p&gt;
&lt;/div&gt;&lt;!-- SC_ON --&gt;",
                "postIframe": "",
                "postImage": "https://b.thumbs.redditmedia.com/5ECi_D2EZxcveWeDzFK8bdMu2pXXeu_oO-cxk0AEtYs.jpg",
                "postThumbnail": "",
                "postTitle": "Weekend Discussion Thread for the Weekend of April 01, 2021",
                "postVideo": "",
                "subreddit": "r/wallstreetbets",
                "time": 1617307633,
                "upvotes": 10,
                "user": "OPINION_IS_UNPOPULAR",
              }
            }
          >
            <div
              className="Post--Content"
            >
              <h3
                className="Post--Title"
                data-read-more="Click Title to Open"
              >
                Weekend Discussion Thread for the Weekend of April 01, 2021
              </h3>
              <figure
                className="Post--Figure"
              >
                <img
                  alt="Weekend Discussion Thread for the Weekend of April 01, 2021"
                  className="Post--Image"
                  loading="lazy"
                  src="https://b.thumbs.redditmedia.com/5ECi_D2EZxcveWeDzFK8bdMu2pXXeu_oO-cxk0AEtYs.jpg"
                />
              </figure>
              <section>
                <div
                  className="Post--Body"
                >
                  <div
                    className="md"
                    key="1"
                  >
                    <p
                      key="0"
                    >
                      Your weekend discussion thread. Please keep the shitposting to a maximum!
                    </p>
                    


                    <p
                      key="2"
                    >
                      Follow 
                      <a
                        href="https://twitter.com/Official_WSB"
                        key="1"
                      >
                        @Official_WSB
                      </a>
                       on Twitter, all other accounts are impersonators.
                    </p>
                    

                  </div>
                </div>
                <div
                  className="Post--Bottom"
                >
                  <div
                    className="Post--Comments"
                  >
                    <BiComment
                      className="comment-icon icon"
                    >
                      <IconBase
                        attr={
                          Object {
                            "viewBox": "0 0 24 24",
                          }
                        }
                        className="comment-icon icon"
                      >
                        <svg
                          className="comment-icon icon"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          strokeWidth="0"
                          style={
                            Object {
                              "color": undefined,
                            }
                          }
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M20,2H4C2.897,2,2,2.897,2,4v18l5.333-4H20c1.103,0,2-0.897,2-2V4C22,2.897,21.103,2,20,2z M20,16H6.667L4,18V4h16V16z"
                            key="0"
                          />
                        </svg>
                      </IconBase>
                    </BiComment>
                    <span
                      className="comments value"
                    >
                      10
                    </span>
                    <span
                      className="label"
                    >
                      comments
                    </span>
                  </div>
                  <BsDot
                    className="dot-icon"
                  >
                    <IconBase
                      attr={
                        Object {
                          "fill": "currentColor",
                          "viewBox": "0 0 16 16",
                        }
                      }
                      className="dot-icon"
                    >
                      <svg
                        className="dot-icon"
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        strokeWidth="0"
                        style={
                          Object {
                            "color": undefined,
                          }
                        }
                        viewBox="0 0 16 16"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          clipRule="evenodd"
                          d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
                          fillRule="evenodd"
                          key="0"
                        />
                      </svg>
                    </IconBase>
                  </BsDot>
                  <div
                    className="Votes Post--Upvotes"
                  >
                    <GoArrowUp
                      className="arrow-icon icon"
                    >
                      <IconBase
                        attr={
                          Object {
                            "viewBox": "0 0 10 16",
                          }
                        }
                        className="arrow-icon icon"
                      >
                        <svg
                          className="arrow-icon icon"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          strokeWidth="0"
                          style={
                            Object {
                              "color": undefined,
                            }
                          }
                          viewBox="0 0 10 16"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M5 3L0 9h3v4h4V9h3L5 3z"
                            fillRule="evenodd"
                            key="0"
                          />
                        </svg>
                      </IconBase>
                    </GoArrowUp>
                    <span
                      className="upvotes value"
                    >
                      10
                    </span>
                    <span
                      className="label"
                    >
                      upvotes
                    </span>
                  </div>
                  <BsDot
                    className="dot-icon"
                  >
                    <IconBase
                      attr={
                        Object {
                          "fill": "currentColor",
                          "viewBox": "0 0 16 16",
                        }
                      }
                      className="dot-icon"
                    >
                      <svg
                        className="dot-icon"
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        strokeWidth="0"
                        style={
                          Object {
                            "color": undefined,
                          }
                        }
                        viewBox="0 0 16 16"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          clipRule="evenodd"
                          d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
                          fillRule="evenodd"
                          key="0"
                        />
                      </svg>
                    </IconBase>
                  </BsDot>
                  <div
                    className="Votes Post--Downvotes"
                  >
                    <GoArrowDown
                      className="arrow-icon icon"
                    >
                      <IconBase
                        attr={
                          Object {
                            "viewBox": "0 0 10 16",
                          }
                        }
                        className="arrow-icon icon"
                      >
                        <svg
                          className="arrow-icon icon"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          strokeWidth="0"
                          style={
                            Object {
                              "color": undefined,
                            }
                          }
                          viewBox="0 0 10 16"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M7 7V3H3v4H0l5 6 5-6H7z"
                            fillRule="evenodd"
                            key="0"
                          />
                        </svg>
                      </IconBase>
                    </GoArrowDown>
                    <span
                      className="downvotes value"
                    >
                      0
                    </span>
                    <span
                      className="label"
                    >
                      downvotes
                    </span>
                  </div>
                </div>
              </section>
            </div>
          </PostContent>
        </div>
      </div>
    </div>
  </Post>
</Provider>
`;
